crates/sp-domains-fraud-proof/src/host_functions.rs

```diff
@@ -98,11 +99,8 @@ where
         domain_id: DomainId,
     ) -> Option<Vec<u8>> {
         let runtime_api = self.consensus_client.runtime_api();
-        let consensus_block_hash = consensus_block_hash.into();
-        let runtime_code = runtime_api
-            .domain_runtime_code(consensus_block_hash, domain_id)
-            .ok()??;
-        let timestamp = runtime_api.timestamp(consensus_block_hash).ok()?;
+        let runtime_code = self.get_domain_runtime_code(consensus_block_hash, domain_id)?;
```

Comment:
I'm mostly thinking about the upgradability of the whole verification. If we make the domain runtime a parameter then the consensus runtime can decide whether the domain runtime code is coming from another host function or from the runtime state (i.e. if we decide to maintain the old domain runtime code through reference counter in the future), but yes I don't have a specific use case for now and maybe overthinking thus no strong opinion. Also WDYT @nazar-pc

Reply:
Honestly I'm not entirely sure what is happening here to comment on it

Reply:
Like whether we should split the host function into some more composable and smaller host functions.

Reply:
Abstractly yes, I was suggesting that before (when I was advocating for more narrow well named calls rather than calls specific to specific fraud proofs), but I'm not sure about this case, how it would interact with something.

Reply:
What @NingLin-P is proposing here is as follows:
When we try to derive `Timestamp` or `SetCode` extrinsics, which by themselves are already two different host calls, in the host function, they fetch the domain runtime from consensus state and then use that domain runtime to derive such extrinsics. 
What he is proposing is to pass the domain runtime code from the host function, then pass that domain runtime code as a parameter to extrinsic derivation instead.

I'm not super excited about that idea since that seems unnecessary as the derivation of the extrinsics are themselves generic enough and I'm worried about the passing an entire domain runtime as a param from runtime through host function, which was anyway fetched from the host function

Reply:
Well, the host function makes runtime API call that still results in interaction with runtime, which also means encoding it and crossing runtime boundary yet another time. It might be more efficient to send it right away, but I'm not sure how much nicer it is or whether it is actually better in any significant way.

Reply:
true, but with this approach it will 
- fetch from runtime 
- pass it back to runtime
- pass from runtime to host function again

Current change results in just fetching from runtime and no passes necessary.
I think what I would like to bring from this discussion to avoid similar discussion in the future is to have a clear definition of how composable do we want our host functions to be.

Reply:
I'm not sure why it needs to do all of that dancing, can't runtime read it from the storage directly before returning? Domain runtime is just a storage item in consensus runtime.

Reply:
> Maybe we may not have that result result yet but if this is important then we should have a seperate disscussion on this

There is no specific use case I can think of for now as mentioned above, but yes I think what I wanted is a discussion about what the host function should be, considering upgradability.

Because now I do see a trend that we putting a lot of code into the host function for efficiency/convenience (mostly the same  `get_fraud_proof_verification_info` host function) while there are only a few types of code that have to be in the host function:
- Getting historical state
- Getting block body (for the bundle)
- Operation not supported in WASM (i.e. `execution_proof_check`)

The domain runtime may not be a good example to discuss, perhaps https://github.com/subspace/subspace/pull/2132/files/c3be85f0c5b0b7e9079166df2c2731587ca690bd#diff-01b339e04054e6a85f7589f9428e0070a58456ddae6eaf0e95a7efb63ffa8ffeR119